### PR TITLE
drivers: clock_control: nrf: Use synthesized LF clock source in tests

### LINKED ISSUE #24784: nRF: Busy wait clock is skewed vs. timer clock
On nRF5 SoCs, the integration uses CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT=y to forward k_busy_wait() calls to a HAL-defined nrfx_coredep_delay_us() function (which presumably works at CPU cycle precision) instead of using the default Zephyr one (which is limited to the 32kHz cycle rate).

Unfortunately, these clocks don't match.  The HAL timer is consistently about .5-.6% too fast, meaning that k_busy_wait() will return earlier than users expect.

This is causing a test failure on at least nrf52840 devices, where tests/kernel/timer/timer_api has a few long delays that it expects to be correct relative to the tick/cycle timer.  This got unmasked by the timer API changes (which fixed a double-round-up in the k_timer_remaining_*() paths), but it seems like the problem is definitely at the platform layer.

The code below does a bunch of busy waits at different delays and compares them to the times reported by the existing timer driver.  It produces the following log on my nrf52840dk_nrf52840.  Note that as the delays increase the error settles down to a consistent asymptote, implying that it's the clock rates that are skewed and not just an offset or aliased conversion somewhere:

(Note also that as I iterate the test, the deltas are jumping around quite a bit to the tune of 5-10 32kHz cycles per half-second busy wait.  300us of jitter seems like **way** too much to me given the clock rates involved.)

```
*** Booting Zephyr OS build zephyr-v2.2.0-2044-g315cba1c40fc  ***
                eff   err
   us    cyc     us   ppm
 ===== ====== ====== =====
     1      0      0 1000000
     2      0      0 1000000
     4      0      0 1000000
     8      0      0 1000000
    16      1     30     0
    32      1     30 500000
    64      2     61 333333
   128      5    152     0
   256      8    244 111111
   512     16    488 58823
  1024     33   1007 29411
  2048     67   2044 14705
  4096    134   4089  7407
  8192    267   8148  7434
 16384    535  16326  3724
 32768   1070  32653  3724
 65536   2141  65338  3258
131072   4275 130462  4656
262144   8542 260681  5587
524288  17080 521240  5820
```

```c
#include <zephyr.h>
#include <stdlib.h>

void main(void)
{
	printk("                eff   err\n");
	printk("   us    cyc     us   ppm\n");
	printk(" ===== ====== ====== =====\n");
	for (int us = 1; us < 1000000; us *= 2) {

		/* Busy Wait */
		int t0 = k_cycle_get_32();
		k_busy_wait(us);
		int t1 = k_cycle_get_32();

		/* Effective us waited: floor, because we want to wait
		 * at least this long and this is the longest duration
		 * we can prove.
		 */
		int eff_us = k_cyc_to_us_floor32(t1 - t0);

		/* Error, in parts per million */
		int exp_cyc = k_us_to_cyc_ceil32(us);
		int err_ppm = (int)(1000000ull
				    * abs((t1 - t0) - exp_cyc) / exp_cyc);

		printk("%6d %6d %6d %5d\n", us, t1-t0, eff_us, err_ppm);
	}
}
```

### PR DESCRIPTION
**Pull request by nordic-krch**
In tests k_busy_wait is used extensively. It  measures delay using
cpu clock source (HF RC). It is less accurate than LF default clock
source (LF XTAL) which leads to tests failure where it is expected
that k_busy_wait(n) will give same (with microseconds accuracy) to
k_sleep(K_MSEC(1000*n)). By chosing LF SYNTH as the source for LF
clock (system clock), it is ensured that HF (cpu) and LF (system)
clocks have the same accuracy and k_sleep and k_busy_wait results
are aligned.

Fixes #24784 for testing.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Review by pabigot — COMMENTED** [review]
I agree this should solve the problem, but it does it by selecting a test configuration that does not represent what applications do.  I'm less sure that's a good idea as it hides the fact that there is, and will always be, a potential difference between these clock domains.

File: drivers/clock_control/Kconfig.nrf [hunk 1]
```diff
@@ -23,7 +23,8 @@ if CLOCK_CONTROL_NRF
 
 choice CLOCK_CONTROL_NRF_SOURCE
 	prompt "32KHz clock source"
-	default CLOCK_CONTROL_NRF_K32SRC_XTAL
+	default CLOCK_CONTROL_NRF_K32SRC_XTAL if !ZTEST
+	default CLOCK_CONTROL_NRF_K32SRC_SYNTH if ZTEST
```
Comment (carlescufi) [inline drivers/clock_control/Kconfig.nrf:27]: @pabigot @andyross @nordic-krch 

Since the assumption that `k_busy_wait()` and `k_sleep()` are driven by the same clock is just not valid for Nordic platfroms (and likely for others as well, present or future), would it be better to introduce a new `FREQUENCY_TOLERANCE` variable that the test uses to extend its tolerance and that can be set to `0` for platforms that use the same clock.
Then you can have something like: `check_timeout_drift()` that uses `FREQUENCY_TOLERANCE` in all tests that combine busy waits with sleep.

Comment (carlescufi) [reply drivers/clock_control/Kconfig.nrf:27]: @andyross could you please weigh in here?

Comment (andyross) [reply drivers/clock_control/Kconfig.nrf:27]: Comments in https://github.com/zephyrproject-rtos/zephyr/issues/24784#issuecomment-628044358 -- I'd be personally happy with this patch, or its converse.

To nitpick: I guess stylewise I'd prefer a choice of name that conveys the implication and not just the details.  The important thing to an app isn't what whether the source is "SYNTH" or "XTAL", it's that the "synth" choice is subject to temperature and semiconductor process skew where the crystal is tuned to more conventional tolerances.

As far as adding code to the test framework, that would work, but it doesn't seem much different than just adding a fudge factor for individual cases.  These kind of long busy waits are pretty rare even in the tests.

**Review by andyross — APPROVED** [review]
But to be clear: I'd be fine with this as-is.

**Review by andyross — CHANGES_REQUESTED** [review]
Actually, thinking on this a little more.  I'm not quite fine with it as-is.  The effect of the actual logic in this patch is to change the default value of a kconfig depending on whether or not ZTEST is set, which means that the test suite isn't going to be testing Zephyr as it is built by applications.

Now, this particular situation is a tiny edge case and not likely to be a problem.  But design-wise that's sort of a no-no.  We should define the kconfig, what it does, how it's used, and have the tests that need it use it just like apps would.
